apiserver/common/leadership.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package common
+
+import (
+	"github.com/juju/errors"
+	"gopkg.in/juju/names.v2"
+
+	"github.com/juju/juju/apiserver/facade"
+	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/juju/core/leadership"
+	"github.com/juju/juju/permission"
+)
+
+// API exposes leadership pinning and unpinning functionality for remote use.
+type LeadershipAPI interface {
+	PinLeadership(params params.PinLeadershipBulkParams) (params.ErrorResults, error)
+	UnpinLeadership(params params.PinLeadershipBulkParams) (params.ErrorResults, error)
+}
+
+// NewFacade creates and returns a new leadership API.
+// This signature is suitable for facade registration.
+func NewLeadershipFacade(ctx facade.Context) (LeadershipAPI, error) {
+	st := ctx.State()
+	model, err := st.Model()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	pinner, err := ctx.LeadershipPinner(model.UUID())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return NewLeadershipAPI(model.ModelTag(), pinner, ctx.Auth())
+}
+
+// NewAPI creates and returns a new leadership API from the input tag,
+// Pinner implementation and facade Authorizer.
+func NewLeadershipAPI(
+	modelTag names.ModelTag, pinner leadership.Pinner, authorizer facade.Authorizer,
+) (LeadershipAPI, error) {
+	return &leadershipAPI{
+		modelTag:   modelTag,
+		pinner:     pinner,
+		authorizer: authorizer,
+	}, nil
+}
+
+type leadershipAPI struct {
+	modelTag   names.ModelTag
+	pinner     leadership.Pinner
+	authorizer facade.Authorizer
+}
+
+// PinLeadership (API) pins the leadership for applications indicated in the
+// input arguments.
+func (a *leadershipAPI) PinLeadership(param params.PinLeadershipBulkParams) (params.ErrorResults, error) {
+	entity := a.authorizer.GetAuthTag()
+	if err := a.checkAccess(entity, permission.WriteAccess); err != nil {
```

Comment:
Is this the write auth check for all purposes?
It seems like Auth check might be different for a Client api vs an Agent api.

Eg Machine has auth for any application that is running on it (but not other apps), and user has auth if it has Write access to the model.
This feels like it is just the client auth.